source/common/router/rds_impl.h

```diff
@@ -138,21 +122,62 @@ class RdsRouteConfigProviderImpl
   void registerInitTarget(Init::Manager& init_manager);
   void runInitializeCallbackIfAny();
 
-  Server::Configuration::FactoryContext& factory_context_;
   std::unique_ptr<Envoy::Config::Subscription<envoy::api::v2::RouteConfiguration>> subscription_;
-  ThreadLocal::SlotPtr tls_;
   std::string config_source_;
+  std::function<void()> initialize_callback_;
   const std::string route_config_name_;
-  absl::optional<LastConfigInfo> config_info_;
   Stats::ScopePtr scope_;
   RdsStats stats_;
-  std::function<void()> initialize_callback_;
   RouteConfigProviderManagerImpl& route_config_provider_manager_;
   const std::string manager_identifier_;
-  envoy::api::v2::RouteConfiguration route_config_proto_;
+  SystemTimeSource& time_source_;
   SystemTime last_updated_;
+  absl::optional<LastConfigInfo> config_info_;
+  envoy::api::v2::RouteConfiguration route_config_proto_;
+  std::unordered_set<RdsRouteConfigProviderImpl*> route_config_providers_;
```

Comment:
Not to store the whole provider object,   define OnUpdate interface,  only store the interface pointers.

Reply:
ditto above, you need pointer to ProviderImpl to get shared_ptr of it.

Reply:
This is what I mean:

class RdsSubscriptionCallback {
  public:
     virtual onConfigUpdate() PURE;
};

Here RdsConfigSubscription uses it as:

 std::unordered_set<RdsSubscriptionCallback*> update_callbacks_;

Reply:
Yes I know what you meant, then you won't be able to do https://github.com/lizan/envoy/blob/6d433fef72eebc37b63acef63fd492cbbb9ff659/source/common/router/rds_impl.cc#L204 without dynamic_cast.

It doesn't make sense to make an interface which is completely implementation detail.